File: classB/17/homework2/task1.hs
import Data.List.Split (splitOn)
import Data.List (transpose)

main :: IO ()
main = do
    csv1 <- readFile "matrix1.csv"
    csv2 <- readFile "matrix2.csv"
    let m1 = csvToMatrix csv1
    let m2 = csvToMatrix csv2
    let result = addMatrix m1 m2
    putStrLn $  if sameSize m1 m2 then
                    if negateMatrix result == transpose result 
                        then "true"
                        else "false"
                else "undefined"

sameSize :: Num a => [[a]] -> [[a]] -> Bool
sameSize m1 m2 = length m1 == length m2 &&
                    map length m1 == map length m2

addMatrix :: Num a => [[a]] -> [[a]] -> [[a]]
addMatrix = zipWith $ zipWith (+)

negateMatrix :: Num a => [[a]] -> [[a]]
negateMatrix = map $ map negate

csvToMatrix :: String -> [[Double]]
csvToMatrix = (map (map read -- read strings to doubles
                        . (splitOn ","))) -- split values
                            . lines -- split lines

File: classB/02/homework2/task1.rb
require "csv"

matrix1 = CSV.read("./matrix1.csv");
matrix2 = CSV.read("./matrix2.csv");

if matrix1.length != matrix2.length || matrix1[0].length != matrix2[0].length
	p 'undefined'
	return
end

rows = matrix1.length
cols = matrix1[0].length
if cols != rows
	p false
	return
end
result = Array.new(rows){Array.new(cols)}

for i  in 0..rows - 1
	for j in 0..cols - 1
		result[i][j] = matrix1[i][j].to_i + matrix2[i][j].to_i
	end
end

for i  in 0..rows - 1
	for j in 0..cols - 1
		if result[i][j] != result[j][i]
			p false
			return
		end
	end
end

p true
